src/callbacks_step/amr.jl

```diff
@@ -528,6 +528,103 @@ function copy_to_quad_iter_volume(info, user_data)
+                                     semi,
+                                     t, iter;
+                                     only_refine = false, only_coarsen = false,
+                                     passive_args = ())
+    @unpack controller, adaptor = amr_callback
+
+    u = wrap_array(u_ode, mesh, equations, dg, cache)
+    lambda = @trixi_timeit timer() "indicator" controller(u, mesh, equations, dg, cache,
+                                                          t = t, iter = iter)
+
+    @boundscheck begin
+        @assert axes(lambda)==(Base.OneTo(ncells(mesh)),) ("Indicator array (axes = $(axes(lambda))) and mesh cells (axes = $(Base.OneTo(ncells(mesh)))) have different axes")
+    end
+
+    # Copy controller value of each quad to the quad's user data storage
+    iter_volume_c = cfunction(copy_to_quad_iter_volume, Val(ndims(mesh)))
+
+    # The pointer to lambda will be interpreted as Ptr{Int} above
+    @assert lambda isa Vector{Int}
+    iterate_p4est(mesh.p4est, lambda; iter_volume_c = iter_volume_c)
+
+    @trixi_timeit timer() "refine" if !only_coarsen
+        # Refine mesh
+        refined_original_cells = @trixi_timeit timer() "mesh" refine!(mesh)
+
+        # Refine solver
+        @trixi_timeit timer() "solver" refine!(u_ode, adaptor, mesh, equations, dg,
+                                               cache, cache_parabolic,
+                                               refined_original_cells)
+        for (p_u_ode, p_mesh, p_equations, p_dg, p_cache) in passive_args
+            @trixi_timeit timer() "passive solver" refine!(p_u_ode, adaptor, p_mesh,
+                                                           p_equations,
+                                                           p_dg, p_cache,
+                                                           refined_original_cells)
+        end
+    else
+        # If there is nothing to refine, create empty array for later use
+        refined_original_cells = Int[]
+    end
+
+    @trixi_timeit timer() "coarsen" if !only_refine
+        # Coarsen mesh
+        coarsened_original_cells = @trixi_timeit timer() "mesh" coarsen!(mesh)
+
+        # coarsen solver
+        @trixi_timeit timer() "solver" coarsen!(u_ode, adaptor, mesh, equations, dg,
+                                                cache, cache_parabolic,
+                                                coarsened_original_cells)
+        for (p_u_ode, p_mesh, p_equations, p_dg, p_cache) in passive_args
+            @trixi_timeit timer() "passive solver" coarsen!(p_u_ode, adaptor, p_mesh,
+                                                            p_equations,
+                                                            p_dg, p_cache,
+                                                            coarsened_original_cells)
+        end
+    else
+        # If there is nothing to coarsen, create empty array for later use
+        coarsened_original_cells = Int[]
+    end
```

Comment:
All of this looks extremely familiar, so I'm assuming it's pretty much identical to other solvers? Would it make sense to extract a function to avoid duplicate code?

Reply:
Absolutely, this is pretty much the same as for the hyperbolic case. We talked about this during one of the Trixi meetings, i.e., code duplication/specializations vs. essentially if-clauses in the functions. The outcome was that we do prioritize simplicity and ease of understanding over a potentially shorter implementation.

Reply:
Hm not sure, this is so high-level that it is probably not affected by that.